File: packages/snow_draw_core/lib/draw/edit/core/edit_session_service.dart
import 'package:meta/meta.dart';

import '../../config/draw_config.dart';
import '../../models/draw_state.dart';
import '../../models/edit_session_id.dart';
import '../../models/interaction_state.dart';
import '../../services/log/log_service.dart';
import '../../services/text/text_metrics_service.dart';
import '../../types/draw_point.dart';
import '../../types/edit_operation_id.dart';
import '../../types/snap_guides.dart';
import '../edit_operation_registry_interface.dart';
import 'edit_error_handler.dart';
import 'edit_modifiers.dart';
import 'edit_operation.dart';
import 'edit_operation_params.dart';
import 'edit_result_unified.dart';

/// Edit action pipeline (route A): keep store-side session handling,
/// but centralize the orchestration into a small, testable service.
@immutable
class EditSessionService {
  EditSessionService({
    required this.editOperations,
    required this.configProvider,
    this.textMetricsService = defaultTextMetricsService,
    LogService? logService,
  }) : _log = logService?.edit;

  factory EditSessionService.fromRegistry(
    EditOperationRegistry registry, {
    required DrawConfig Function() configProvider,
    TextMetricsService textMetricsService = defaultTextMetricsService,
    LogService? logService,
  }) => EditSessionService(
    editOperations: registry,
    configProvider: configProvider,
    textMetricsService: textMetricsService,
    logService: logService,
  );
  final EditOperationRegistry editOperations;
  final DrawConfig Function() configProvider;
  final TextMetricsService textMetricsService;
  final ModuleLogger? _log;

  // Session API.
  EditOutcome start({
    required DrawState state,
    required EditOperationId operationId,
    required DrawPoint position,
    required EditOperationParams params,
    required EditSessionId sessionId,
  }) {
    if (!state.domain.selection.hasSelection) {
      return (
        state: state,
        failureReason: EditFailureReason.noSelection,
        operationId: operationId,
      );
    }

    final operation = editOperations.getOperation(operationId);
    if (operation == null) {
      return (
        state: state,
        failureReason: EditFailureReason.unknownOperationId,
        operationId: operationId,
      );
    }

    return EditErrorHandler.runWithErrorHandling(
      state: state,
      keepStateOnFailure: true,
      fallbackOperationId: operationId,
      operationName: 'startEdit',
      log: _log,
      operation: () => _performStart(
        state: state,
        operation: operation,
        operationId: operationId,
        position: position,
        params: params,
        sessionId: sessionId,
      ),
    );
  }

  EditOutcome update({
    required DrawState state,
    required DrawPoint currentPosition,
    EditModifiers modifiers = const EditModifiers(),
  }) => _withRestoredSession(
    state: state,
    operationName: 'updateEdit',
    validateVersions: true,
    action: (restored) => _performUpdate(
      state: state,
      operation: restored.operation,
      editingState: restored.editingState,
      currentPosition: currentPosition,
      modifiers: modifiers,
    ),
  );

  EditOutcome finish({required DrawState state}) => _withRestoredSession(
    state: state,
    operationName: 'finishEdit',
    validateVersions: true,
    action: (restored) => _performFinish(
      state: state,
      operation: restored.operation,
      editingState: restored.editingState,
    ),
  );

  EditOutcome cancel({required DrawState state}) => _withRestoredSession(
    state: state,
    operationName: 'cancelEdit',
    action: (restored) => _performCancel(
      state: state,
      operation: restored.operation,
      editingState: restored.editingState,
    ),
  );

  EditOutcome _withRestoredSession({
    required DrawState state,
    required String operationName,
    required EditOutcome Function(
      ({EditOperation operation, EditingState editingState}) restored,
    )
    action,
    bool validateVersions = false,
  }) {
    final restoration = _restoreSession(
      state,
      validateVersions: validateVersions,
    );
    if (restoration.failureReason case final reason?) {
      return EditErrorHandler.createFailure(state: state, reason: reason);
    }

    final restored = restoration.session!;
    return EditErrorHandler.runWithErrorHandling(
      state: state,
      operationName: operationName,
      log: _log,
      fallbackOperationId: restored.editingState.operationId,
      operation: () => action(restored),
    );
  }

  EditOutcome _successOutcome({
    required DrawState state,
    required EditOperationId operationId,
  }) => (state: state, failureReason: null, operationId: operationId);

  T _runWithTextMetrics<T>(T Function() action) =>
      runWithScopedTextMetricsService(
        textMetricsService: textMetricsService,
        action: action,
      );

  EditOutcome _performStart({
    required DrawState state,
    required EditOperation operation,
    required EditOperationId operationId,
    required DrawPoint position,
    required EditOperationParams params,
    required EditSessionId sessionId,
  }) {
    final session = _createSession(
      operation: operation,
      operationId: operationId,
      state: state,
      position: position,
      params: params,
      sessionId: sessionId,
    );
    final nextApplication = state.application.copyWith(interaction: session);
    return _successOutcome(
      state: state.copyWith(application: nextApplication),
      operationId: operationId,
    );
  }

  EditOutcome _performUpdate({
    required DrawState state,
    required EditOperation operation,
    required EditingState editingState,
    required DrawPoint currentPosition,
    required EditModifiers modifiers,
  }) {
    _log?.trace('Edit session updated', {'operationId': operation.id});
    final updated = _runWithTextMetrics(
      () => operation.update(
        state: state,
        context: editingState.context,
        transform: editingState.currentTransform,
        currentPosition: currentPosition,
        modifiers: modifiers,
        config: configProvider(),
      ),
    );

    final transformUnchanged =
        updated.transform == editingState.currentTransform;
    final guidesUnchanged = snapGuideListEquals(
      updated.snapGuides,
      editingState.snapGuides,
    );
    if (transformUnchanged && guidesUnchanged) {
      return _successOutcome(
        state: state,
        operationId: editingState.operationId,
      );
    }

    return _successOutcome(
      state: state.copyWith(
        application: state.application.copyWith(
          interaction: editingState.withTransform(
            updated.transform,
            guides: updated.snapGuides,
          ),
        ),
      ),
      operationId: editingState.operationId,
    );
  }

  EditOutcome _performFinish({
    required DrawState state,
    required EditOperation operation,
    required EditingState editingState,
  }) {
    _log?.info('Edit session finished', {'operationId': operation.id});
    return _successOutcome(
      state: _runWithTextMetrics(
        () => operation.finish(
          state: state,
          context: editingState.context,
          transform: editingState.currentTransform,
        ),
      ),
      operationId: editingState.operationId,
    );
  }

  EditOutcome _performCancel({
    required DrawState state,
    required EditOperation operation,
    required EditingState editingState,
  }) {
    _log?.info('Edit session cancelled', {'operationId': operation.id});
    return _successOutcome(
      state: operation.cancel(state: state),
      operationId: editingState.operationId,
    );
  }

  EditingState _createSession({
    required EditOperation operation,
    required EditOperationId operationId,
    required DrawState state,
    required DrawPoint position,
    required EditOperationParams params,
    required EditSessionId sessionId,
  }) {
    _log?.info('Edit session created', {
      'operationId': operationId,
      'params': params.runtimeType.toString(),
    });
    final sessionData = _runWithTextMetrics(() {
      final context = operation.createContext(
        state: state,
        position: position,
        params: params,
      );
      final transform = operation.initialTransform(
        state: state,
        context: context,
        startPosition: position,
      );
      return (context: context, transform: transform);
    });
    return EditingState(
      operationId: operationId,
      sessionId: sessionId,
      context: sessionData.context,
      currentTransform: sessionData.transform,
    );
  }

  ({
    ({EditOperation operation, EditingState editingState})? session,
    EditFailureReason? failureReason,
  })
  _restoreSession(DrawState state, {bool validateVersions = false}) {
    final interaction = state.application.interaction;
    if (interaction is! EditingState) {
      _log?.error('Edit session restore failed', null, null, {
        'reason': 'not_editing',
      });
      return (session: null, failureReason: EditFailureReason.notEditing);
    }

    final operation = editOperations.getOperation(interaction.operationId);
    if (operation == null) {
      _log?.error('Edit session restore failed', null, null, {
        'operationId': interaction.operationId,
        'reason': 'unknown_operation',
      });
      return (
        session: null,
        failureReason: EditFailureReason.unknownOperationId,
      );
    }

    if (validateVersions) {
      final versionConflict = _resolveVersionConflict(
        editingState: interaction,
        currentState: state,
      );
      if (versionConflict case final reason?) {
        return (session: null, failureReason: reason);
      }
    }

    _log?.trace('Edit session restored', {
      'operationId': interaction.operationId,
      'sessionId': interaction.sessionId,
    });
    return (
      session: (operation: operation, editingState: interaction),
      failureReason: null,
    );
  }

  EditFailureReason? _resolveVersionConflict({
    required EditingState editingState,
    required DrawState currentState,
  }) {
    final context = editingState.context;
    final currentSelectionVersion =
        currentState.domain.selection.selectionVersion;
    if (context.selectionVersion != currentSelectionVersion) {
      _log?.warning('Edit session version conflict', {
        'operationId': editingState.operationId,
        'type': 'selection',
        'expected': context.selectionVersion,
        'actual': currentSelectionVersion,
      });
      return EditFailureReason.selectionChanged;
    }

    final currentElementsVersion = currentState.domain.document.elementsVersion;
    if (context.elementsVersion != currentElementsVersion) {
      _log?.warning('Edit session version conflict', {
        'operationId': editingState.operationId,
        'type': 'elements',
        'expected': context.elementsVersion,
        'actual': currentElementsVersion,
      });
      return EditFailureReason.elementsChanged;
    }
    return null;
  }
}

File: packages/snow_draw_rust_engine/src/draw/edit/core/edit_session_service.rs
#![allow(dead_code)]

use std::sync::Arc;

use crate::draw::config::draw_config::DrawConfig;
use crate::draw::edit::edit_operations::DefaultEditOperationRegistry;
use crate::draw::models::draw_state::DrawState;
use crate::draw::models::edit_session_id::EditSessionId;
use crate::draw::models::interaction_state::{EditingState, InteractionState};
use crate::draw::services::log::log_service::{LogData, ModuleLogger};
use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::edit_context::{
    default_text_metrics_service, EditContext, TextMetricsService,
};
use crate::draw::types::edit_operation_id::EditOperationId;
use crate::draw::types::snap_guides::{snap_guide_list_equals, SnapGuide};

use super::edit_error_handler::{EditErrorHandler, EditFailureReason, EditOutcome};
use super::edit_modifiers::EditModifiers;
use super::edit_operation::{EditOperation, EditOperationParams};

#[derive(Clone)]
struct RestoredSession {
    operation: Arc<dyn EditOperation>,
    editing_state: EditingState,
}

#[derive(Default)]
struct SessionRestoreResult {
    session: Option<RestoredSession>,
    failure_reason: Option<EditFailureReason>,
    fallback_operation_id: Option<EditOperationId>,
}

impl SessionRestoreResult {
    fn success(session: RestoredSession) -> Self {
        Self {
            session: Some(session),
            failure_reason: None,
            fallback_operation_id: None,
        }
    }

    fn failure(reason: EditFailureReason, fallback_operation_id: Option<EditOperationId>) -> Self {
        Self {
            session: None,
            failure_reason: Some(reason),
            fallback_operation_id,
        }
    }
}

/// Edit action pipeline.
pub struct EditSessionService {
    pub edit_operations: DefaultEditOperationRegistry,
    config_provider: Arc<dyn Fn() -> DrawConfig + Send + Sync>,
    text_metrics_service: Arc<dyn TextMetricsService>,
    log: Option<ModuleLogger>,
}

impl EditSessionService {
    pub fn new(
        edit_operations: DefaultEditOperationRegistry,
        config_provider: impl Fn() -> DrawConfig + Send + Sync + 'static,
        text_metrics_service: Option<Arc<dyn TextMetricsService>>,
        log: Option<ModuleLogger>,
    ) -> Self {
        Self {
            edit_operations,
            config_provider: Arc::new(config_provider),
            text_metrics_service: text_metrics_service.unwrap_or_else(default_text_metrics_service),
            log,
        }
    }

    pub fn from_registry(
        registry: DefaultEditOperationRegistry,
        config_provider: impl Fn() -> DrawConfig + Send + Sync + 'static,
        text_metrics_service: Option<Arc<dyn TextMetricsService>>,
        log: Option<ModuleLogger>,
    ) -> Self {
        Self::new(registry, config_provider, text_metrics_service, log)
    }

    pub fn start(
        &self,
        state: &DrawState,
        operation_id: EditOperationId,
        position: DrawPoint,
        params: &EditOperationParams,
        session_id: EditSessionId,
    ) -> EditOutcome {
        if !state.domain.selection.has_selection() {
            return EditErrorHandler::create_failure(
                state,
                EditFailureReason::NoSelection,
                Some(operation_id),
                true,
            );
        }

        let Some(operation) = self.edit_operations.get_operation(operation_id).cloned() else {
            return EditErrorHandler::create_failure(
                state,
                EditFailureReason::UnknownOperationId,
                Some(operation_id),
                true,
            );
        };

        EditErrorHandler::run_with_error_handling(
            state,
            || self.perform_start(state, operation, operation_id, position, params, session_id),
            true,
            Some(operation_id),
            Some("startEdit"),
            self.log.as_ref(),
        )
    }

    pub fn update(
        &self,
        state: &DrawState,
        current_position: DrawPoint,
        modifiers: EditModifiers,
    ) -> EditOutcome {
        self.with_restored_session(state, "updateEdit", true, |restored| {
            self.perform_update(state, restored, current_position, modifiers)
        })
    }

    pub fn finish(&self, state: &DrawState) -> EditOutcome {
        self.with_restored_session(state, "finishEdit", true, |restored| {
            self.perform_finish(state, restored)
        })
    }

    pub fn cancel(&self, state: &DrawState) -> EditOutcome {
        self.with_restored_session(state, "cancelEdit", false, |restored| {
            self.perform_cancel(state, restored)
        })
    }

    fn with_restored_session<F>(
        &self,
        state: &DrawState,
        operation_name: &str,
        validate_versions: bool,
        action: F,
    ) -> EditOutcome
    where
        F: FnOnce(RestoredSession) -> EditOutcome,
    {
        let restoration = self.restore_session(state, validate_versions);
        if let Some(reason) = restoration.failure_reason {
            return EditErrorHandler::create_failure(
                state,
                reason,
                restoration.fallback_operation_id,
                false,
            );
        }

        let Some(restored) = restoration.session else {
            return EditErrorHandler::create_failure(
                state,
                EditFailureReason::NotEditing,
                None,
                false,
            );
        };

        let fallback_operation_id = Some(restored.editing_state.operation_id);
        EditErrorHandler::run_with_error_handling(
            state,
            || action(restored),
            false,
            fallback_operation_id,
            Some(operation_name),
            self.log.as_ref(),
        )
    }

    fn perform_start(
        &self,
        state: &DrawState,
        operation: Arc<dyn EditOperation>,
        operation_id: EditOperationId,
        position: DrawPoint,
        params: &EditOperationParams,
        session_id: EditSessionId,
    ) -> EditOutcome {
        let session = self.create_session(
            operation.as_ref(),
            operation_id,
            state,
            position,
            params,
            session_id,
        );

        let next_application =
            state
                .application
                .copy_with(None, Some(InteractionState::Editing(session)), None);
        let next_state = state.copy_with(None, Some(next_application));
        self.success_outcome(next_state, operation_id)
    }

    fn perform_update(
        &self,
        state: &DrawState,
        restored: RestoredSession,
        current_position: DrawPoint,
        modifiers: EditModifiers,
    ) -> EditOutcome {
        let editing_state = restored.editing_state;
        self.log_trace(
            "Edit session updated",
            Some(editing_state.operation_id),
            None,
        );
        let config = (self.config_provider)();
        let updated = restored.operation.update(
            state,
            &editing_state.context,
            &editing_state.current_transform,
            current_position,
            modifiers,
            &config,
        );

        let transform_unchanged = updated.transform == editing_state.current_transform;
        let guides_unchanged = snap_guide_list_equals(
            updated.snap_guides.as_slice(),
            editing_state.snap_guides.as_slice(),
        );

        if transform_unchanged && guides_unchanged {
            return self.success_outcome(state.clone(), editing_state.operation_id);
        }

        let next_editing =
            editing_state.with_transform(updated.transform, Some(updated.snap_guides));
        let next_application =
            state
                .application
                .copy_with(None, Some(InteractionState::Editing(next_editing)), None);
        let next_state = state.copy_with(None, Some(next_application));
        self.success_outcome(next_state, editing_state.operation_id)
    }

    fn perform_finish(&self, state: &DrawState, restored: RestoredSession) -> EditOutcome {
        let operation_id = restored.editing_state.operation_id;
        self.log_info("Edit session finished", Some(operation_id), None);
        let finished_state = restored.operation.finish(
            state,
            &restored.editing_state.context,
            &restored.editing_state.current_transform,
        );
        self.success_outcome(finished_state, operation_id)
    }

    fn perform_cancel(&self, state: &DrawState, restored: RestoredSession) -> EditOutcome {
        let operation_id = restored.editing_state.operation_id;
        self.log_info("Edit session cancelled", Some(operation_id), None);
        let cancelled_state = restored.operation.cancel(state);
        self.success_outcome(cancelled_state, operation_id)
    }

    fn create_session(
        &self,
        operation: &dyn EditOperation,
        operation_id: EditOperationId,
        state: &DrawState,
        position: DrawPoint,
        params: &EditOperationParams,
        session_id: EditSessionId,
    ) -> EditingState {
        let mut data = LogData::new();
        data.insert("operationId".to_owned(), operation_id.to_owned());
        data.insert(
            "params".to_owned(),
            match params {
                EditOperationParams::Move(_) => "MoveOperationParams",
                EditOperationParams::Resize(_) => "ResizeOperationParams",
                EditOperationParams::Rotate(_) => "RotateOperationParams",
                EditOperationParams::ConnectorPoint(_) => "ConnectorPointOperationParams",
            }
            .to_owned(),
        );
        self.log_info("Edit session created", Some(operation_id), Some(data));
        let created_context = operation.create_context(state, position, params);
        let context = self.attach_text_metrics_service(operation, created_context);
        let transform = operation.initial_transform(state, &context, position);

        EditingState::new(
            operation_id,
            session_id,
            context,
            transform,
            Vec::<SnapGuide>::new(),
        )
    }

    fn attach_text_metrics_service(
        &self,
        operation: &dyn EditOperation,
        context: EditContext,
    ) -> EditContext {
        operation.attach_text_metrics_service(&context, self.text_metrics_service.clone());
        context
    }

    fn restore_session(&self, state: &DrawState, validate_versions: bool) -> SessionRestoreResult {
        let InteractionState::Editing(editing_state) = &state.application.interaction else {
            let mut data = LogData::new();
            data.insert("reason".to_owned(), "not_editing".to_owned());
            self.log_error("Edit session restore failed", None, Some(data));
            return SessionRestoreResult::failure(EditFailureReason::NotEditing, None);
        };

        let Some(operation) = self
            .edit_operations
            .get_operation(editing_state.operation_id)
            .cloned()
        else {
            let mut data = LogData::new();
            data.insert("reason".to_owned(), "unknown_operation".to_owned());
            self.log_error(
                "Edit session restore failed",
                Some(editing_state.operation_id),
                Some(data),
            );
            return SessionRestoreResult::failure(
                EditFailureReason::UnknownOperationId,
                Some(editing_state.operation_id),
            );
        };

        if validate_versions {
            if let Some(reason) = self.resolve_version_conflict(editing_state, state) {
                return SessionRestoreResult::failure(reason, Some(editing_state.operation_id));
            }
        }

        let mut data = LogData::new();
        data.insert("sessionId".to_owned(), editing_state.session_id.clone());
        self.log_trace(
            "Edit session restored",
            Some(editing_state.operation_id),
            Some(data),
        );
        SessionRestoreResult::success(RestoredSession {
            operation,
            editing_state: editing_state.clone(),
        })
    }

    fn resolve_version_conflict(
        &self,
        editing_state: &EditingState,
        current_state: &DrawState,
    ) -> Option<EditFailureReason> {
        let current_selection_version = current_state.domain.selection.selection_version as i64;
        if editing_state.context.selection_version != current_selection_version {
            let mut data = LogData::new();
            data.insert("type".to_owned(), "selection".to_owned());
            data.insert(
                "expected".to_owned(),
                editing_state.context.selection_version.to_string(),
            );
            data.insert("actual".to_owned(), current_selection_version.to_string());
            self.log_warning(
                "Edit session version conflict",
                Some(editing_state.operation_id),
                Some(data),
            );
            return Some(EditFailureReason::SelectionChanged);
        }

        let current_elements_version = current_state.domain.document.elements_version;
        if editing_state.context.elements_version != current_elements_version {
            let mut data = LogData::new();
            data.insert("type".to_owned(), "elements".to_owned());
            data.insert(
                "expected".to_owned(),
                editing_state.context.elements_version.to_string(),
            );
            data.insert("actual".to_owned(), current_elements_version.to_string());
            self.log_warning(
                "Edit session version conflict",
                Some(editing_state.operation_id),
                Some(data),
            );
            return Some(EditFailureReason::ElementsChanged);
        }

        None
    }

    fn success_outcome(&self, state: DrawState, operation_id: EditOperationId) -> EditOutcome {
        EditOutcome::new(state, None, Some(operation_id))
    }

    fn log_trace(
        &self,
        message: &str,
        operation_id: Option<EditOperationId>,
        data: Option<LogData>,
    ) {
        let Some(log) = self.log.as_ref() else {
            return;
        };

        let resolved_data = Self::with_operation_id(data, operation_id);
        log.trace(message, Some(&resolved_data));
    }

    fn log_info(
        &self,
        message: &str,
        operation_id: Option<EditOperationId>,
        data: Option<LogData>,
    ) {
        let Some(log) = self.log.as_ref() else {
            return;
        };

        let resolved_data = Self::with_operation_id(data, operation_id);
        log.info(message, Some(&resolved_data));
    }

    fn log_warning(
        &self,
        message: &str,
        operation_id: Option<EditOperationId>,
        data: Option<LogData>,
    ) {
        let Some(log) = self.log.as_ref() else {
            return;
        };

        let resolved_data = Self::with_operation_id(data, operation_id);
        log.warning(message, Some(&resolved_data));
    }

    fn log_error(
        &self,
        message: &str,
        operation_id: Option<EditOperationId>,
        data: Option<LogData>,
    ) {
        let Some(log) = self.log.as_ref() else {
            return;
        };

        let resolved_data = Self::with_operation_id(data, operation_id);
        log.error(message, None, None, Some(&resolved_data));
    }

    fn with_operation_id(data: Option<LogData>, operation_id: Option<EditOperationId>) -> LogData {
        let mut resolved_data = data.unwrap_or_default();
        if let Some(operation_id) = operation_id {
            resolved_data
                .entry("operationId".to_owned())
                .or_insert_with(|| operation_id.to_owned());
        }
        resolved_data
    }
}
